mentat/language_server.py

```diff
@@ -0,0 +1,253 @@
+            logger.info("The language server has stopped")
+            self._stop_task = None
+
+        self._stop_task = self.loop.create_task(self._stop())
+        self._stop_task.add_done_callback(callback)
+
+        return self._stop_task
+
+
+server = LanguageServer(host="127.0.0.1", port=7798, exit_on_lost_connection=False)
+
+
+@server.feature(lsp.INITIALIZE)
+async def initialize(ls: LanguageServer, params: lsp.InitializeParams):
+    if params.root_path is None:
+        logger.debug("No root path provided, using home directory")
+        ls.cwd = Path.home()
+    else:
+        ls.cwd = Path(params.root_path)
+
+
+@server.feature("mentat/createSession")
+async def create_session(ls: LanguageServer, params: lsp.InitializeParams):
+    assert ls.cwd is not None
+
+    ls.session = Session(cwd=ls.cwd)
+    ls.session.start()
+
+    async def handle_session_stream():
+        if ls.session is None:
+            return
+        async for message in ls.session.stream.listen():
+            logger.debug(f"Received Session Message: {message.data} {message.extra}")
+            ls_message = LanguageServerMessage(type="notification", method="mentat/serverMessage", data=message)
+            ls.send_notification("mentat/serverMessage", ls_message.model_dump(mode="json"))
+
+    async def handle_input_requests():
+        if ls.session is None:
+            return
+        while True:
+            input_request_message = await ls.session.stream.recv("input_request")
+            print("Got input request:", input_request_message)
+            ls.session_input_request_id = input_request_message.id
+            ls_message = LanguageServerMessage(
+                type="notification",
+                method="mentat/inputRequest",
+                data=input_request_message,
+            )
+            ls.send_notification("mentat/inputRequest", ls_message.model_dump(mode="json"))
+
+    ls.handle_session_stream_task = asyncio.create_task(handle_session_stream())
+    ls.handle_input_requests_task = asyncio.create_task(handle_input_requests())
+
+    print("Waiting for input request")
+    input_request_message = await ls.session.stream.recv("input_request")
+    ls.session_input_request_id = input_request_message.id
+
+    print("Got input request:", input_request_message)
+
+
+@server.feature(lsp.SHUTDOWN)
+async def shutdown(ls: LanguageServer):
+    if ls.session is None:
+        return
+    ls.session.stream.send(None, channel="session_exit")
+
+
+@server.feature(lsp.TEXT_DOCUMENT_DID_OPEN)
+async def handle_text_document_did_open(ls: LanguageServer, params: lsp.DidOpenTextDocumentParams):
+    if ls.session is None:
+        return
+    # print("Got params:", params)
+
+
+class Message(NamedTuple):
+    data: Any
+
+
+@server.feature("mentat/clientMessage")
+async def handle_client_message(ls: LanguageServer, message: Any):
+    if ls.session is None:
+        return
+    if ls.session_input_request_id is None:
```

Comment:
This shouldn't care about the session input request id, right? Shouldn't it just forward all received messages from the vscode extension to the stream?